pkg/reconciler/v1alpha1/taskrun/resources/cloudevent.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+	"encoding/json"
+  "errors"
+  "fmt"
+
+	"github.com/cloudevents/sdk-go/pkg/cloudevents"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/client"
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+  "github.com/knative/eventing-sources/pkg/kncloudevents"
+  "github.com/knative/pkg/apis"
+  "go.uber.org/zap"
+
+  "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// TektonEventType holds the types of cloud events sent by Tekton
+type TektonEventType string
+
+const(
+  // TektonTaskRunUnknown is sent for TaskRuns with "ConditionSucceeded" "Unknown"
+  TektonTaskRunUnknown    TektonEventType = "dev.tekton.event.task.unknown"
+  // TektonTaskRunSuccessful is sent for TaskRuns with "ConditionSucceeded" "True"
+  TektonTaskRunSuccessful TektonEventType = "dev.tekton.event.task.successful"
+  // TektonTaskRunFailed is sent for TaskRuns with "ConditionSucceeded" "False"
+  TektonTaskRunFailed     TektonEventType = "dev.tekton.event.task.failed"
+)
+
+type ClientFactory func(...string) (client.Client, error)
+
+// SendCloudEvent sends a Cloud Event to the specified SinkURI
+func SendCloudEvent(sinkURI, eventID, eventSourceURI string, data []byte, eventType TektonEventType, logger *zap.SugaredLogger, cloudEventClientFactory ClientFactory) (cloudevents.Event, error) {
+	var event cloudevents.Event
+  // Setup the cloudevent client
+	cloudEventClient, err := cloudEventClientFactory(sinkURI)
+	if err != nil {
+		logger.Errorf("Error creating the cloud-event client: %s", err)
+    return event, err
+	}
+
+	cloudEventSource := types.ParseURLRef(eventSourceURI)
+	if cloudEventSource == nil {
+		logger.Errorf("Invalid eventSourceURI: %s", eventSourceURI)
+		return event, fmt.Errorf("Invalid eventSourceURI: %s", eventSourceURI)
+	}
+
+	event = cloudevents.Event{
+		Context: cloudevents.EventContextV02{
+			ID:         eventID,
+			Type:       string(eventType),
+			Source:     *cloudEventSource,
+			Extensions: nil,
+		}.AsV02(),
+		Data: data,
+	}
```

Comment:
Where is the "Time" set?

Reply:
It is not set, all the timestamps from the TaskRun are available in the TaskRunStatus which is sent in the body, so I did not bother setting a timestamp there. What would you set, the time that the event is created or the completion time of the TaskRun?

Reply:
Either is fine.  The event listener may sort or filter the events based of the time.  So I believe that it must be set.

Reply:
Ack, I'll put the TaskRun completion status then.